packages/design/src/FormManager/FormEdit/components/common/PatternEditActions.tsx

```diff
@@ -10,9 +10,36 @@ type PatternEditActionsProps = PropsWithChildren<{
 export const PatternEditActions = ({ children }: PatternEditActionsProps) => {
   children;
   const context = useFormManagerStore(state => state.context);
+  const focusPatternId = useFormManagerStore(state => state.focus?.pattern.id);
   const { deleteSelectedPattern } = useFormManagerStore(state => ({
     deleteSelectedPattern: state.deleteSelectedPattern,
   }));
+  const { copyPattern } = useFormManagerStore(state => ({
+    copyPattern: state.copyPattern,
+  }));
+  const pages = useFormManagerStore(state =>
+    Object.values(state.session.form.patterns).filter(p => p.type === 'page')
+  );
+  const fieldsets = useFormManagerStore(state =>
+    Object.values(state.session.form.patterns).filter(
+      p => p.type === 'fieldset'
+    )
+  );
+  const currentPageIndex = pages.findIndex(page =>
+    page.data.patterns.includes(focusPatternId || '')
+  );
+  const currentFieldsetIndex = fieldsets.findIndex(fieldset =>
+    fieldset.data.patterns.includes(focusPatternId)
+  );
+  const sourcePagePatternId = pages[currentPageIndex]?.id;
+  const sourceFieldsetPatternId = fieldsets[currentFieldsetIndex]?.id;
+  const handleCopyPattern = () => {
+    if (sourcePagePatternId && focusPatternId) {
+      copyPattern(sourcePagePatternId, focusPatternId);
+    } else if (sourceFieldsetPatternId && focusPatternId) {
+      copyPattern(sourceFieldsetPatternId, focusPatternId);
+    }
+  };
```

Comment:
Could this logic be moved into the `copyPattern` function, so we can keep the React component free of complexity?

Reply:
Hi Dan. As we discussed, I wasn't able to move the entire logic into the copyPattern function because the argument values (parentPagePattern ID and fieldsetPattern ID) vary depending on which pattern is clicked. However, I did move as many of the copyPattern related variables to the handleCopyPattern function as possible. Unfortunately, the variables that utilize useFormManagerStore couldn’t be moved due to an Invalid Hook Call error.

Reply:
The form manager store values can be read directly from the `copyPattern` function, so most of this can be moved. We can revisit this later.